doom/config.org
#+TITLE: DOOM Emacs Configuration
#+PROPERTY: header-args :tangle config.el

* Meta

Documentation: https://github.com/hlissner/doom-emacs/blob/develop/modules/config/literate/README.org
Inspired by https://raw.githubusercontent.com/daviwil/dotfiles/master/Emacs.org
Examples:
- https://github.com/hlissner/doom-emacs-private/blob/master/config.el

Tangle with command 'org-babel-tangle'

** TODO Update Table of Contents in this file
** TODO Autosave on eval
** TODO out-org
** TODO highlight occurences

* Table of Contents
:PROPERTIES:
:TOC:      :include all :ignore this
:END:
:CONTENTS:
:END:
* General Configuration

** Doom

#+BEGIN_SRC elisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;;           DO NOT MODIFY!
;;       tangeled from config.org

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Witoslaw Koczewski"
      user-mail-address "wi@koczewski.de")

(setq doom-localleader-key ",")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.
#+END_SRC

** Auto-customizations

By default changes made via a customisation interface are added to init.el.
I prefer the idea of using a separate file for this. We just need to change a
setting, and load it if it exists.

#+BEGIN_SRC elisp

(print "[config.org] Auto-customizations")

(setq-default custom-file (expand-file-name "custom.el" doom-private-dir))
(when (file-exists-p custom-file)
  (load custom-file))

 #+END_SRC

** Better Defaults

#+BEGIN_SRC elisp

(setq-default
 ;; ad-redefinition-action 'accept         ; Silence warnings for redefinition
 auto-save-list-file-prefix nil         ; Prevent tracking for auto-saves
 cursor-in-non-selected-windows nil     ; Hide the cursor in inactive windows
 cursor-type '(hbar . 2)                ; Underline-shaped cursor
 custom-unlispify-menu-entries nil      ; Prefer kebab-case for titles
 custom-unlispify-tag-names nil         ; Prefer kebab-case for symbols
 delete-by-moving-to-trash t            ; Delete files to trash
 fill-column 80                         ; Set width for automatic line breaks
 gc-cons-threshold (* 8 1024 1024)      ; We're not using Game Boys anymore
 help-window-select t                   ; Focus new help windows when opened
 indent-tabs-mode nil                   ; Stop using tabs to indent
 inhibit-startup-screen t               ; Disable start-up screen
 initial-scratch-message ""             ; Empty the initial *scratch* buffer
 mouse-yank-at-point t                  ; Yank at point rather than pointer
 read-process-output-max (* 1024 1024)  ; Increase read size per process
 recenter-positions '(5 top bottom)     ; Set re-centering positions
 scroll-conservatively 101              ; Avoid recentering when scrolling far
 scroll-margin 12                       ; Add a margin when scrolling vertically
 select-enable-clipboard t              ; Merge system's and Emacs' clipboard
 scroll-preserve-screen-position 't
 sentence-end-double-space nil          ; Use a single space after dots
 ;; show-help-function nil                 ; Disable help text everywhere
 tab-always-indent 'complete            ; Tab indents first then tries completions
 tab-width 2                            ; Smaller width for tab characters
 uniquify-buffer-name-style 'forward    ; Uniquify buffer names
 ;; warning-minimum-level :error           ; Skip warning buffers
 window-combination-resize t            ; Resize windows proportionally
 x-stretch-cursor t)                    ; Stretch cursor to the glyph width

;; (blink-cursor-mode 0)                   ; Prefer a still cursor
(delete-selection-mode 1)               ; Replace region when inserting text
(fset 'yes-or-no-p 'y-or-n-p)           ; Replace yes/no prompts with y/n
(global-subword-mode 1)                 ; Iterate through CamelCase words
(mouse-avoidance-mode 'exile)           ; Avoid collision of mouse with point
(put 'downcase-region 'disabled nil)    ; Enable downcase-region
(put 'upcase-region 'disabled nil)      ; Enable upcase-region
(set-default-coding-systems 'utf-8)     ; Default to utf-8 encoding

;; Change a few indenting behaviors.
(put 'add-function 'lisp-indent-function 2)
(put 'advice-add 'lisp-indent-function 2)
(put 'plist-put 'lisp-indent-function 2)

#+END_SRC

** Line Numbers

This determines the style of line numbers in effect. If set to `nil', line
numbers are disabled. For relative line numbers, set this to `relative'.

#+BEGIN_SRC elisp

(setq display-line-numbers-type nil)

#+END_SRC

** Theme and Fonts

There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:

#+BEGIN_SRC elisp

(setq doom-theme 'doom-one)
;; (setq doom-theme 'doom-one-light)
;; (setq doom-theme 'doom-vibrant)
;; (setq doom-theme 'doom-gruvbox)
;; (setq doom-theme 'doom-dracula)

;; test ->

(setq doom-font (font-spec :family "Fira Code" :size 14)
      doom-variable-pitch-font (font-spec :family "Ubuntu" :size 15)
      ivy-posframe-font (font-spec :family "Fira Code" :size 17))

#+END_SRC

** Exit without Confirmation

#+BEGIN_SRC elisp

(setq confirm-kill-emacs nil)

#+END_SRC

** Parens

#+BEGIN_SRC elisp

(add-hook 'smartparens-enabled-hook #'smartparens-strict-mode)
(add-hook 'smartparens-enabled-hook #'evil-smartparens-mode)
(add-hook 'smartparens-enabled-hook #'evil-cleverparens-mode)

(map! :map global-map
      :mode evil-cleverparens-mode
      :n "M-l" #'evil-cp->)
(map! :map global-map
      :mode evil-cleverparens-mode
      :n "M-h" #'evil-cp-<)

#+END_SRC

** Keyboard Bindings

*** ESC cancels all

#+begin_src elisp

(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

#+end_src

*** SPC SPC lists commands

#+begin_src elisp

(map! :leader
      :desc "List commands"
      "SPC"  #'execute-extended-command)

#+end_src

*** Navigation

#+BEGIN_SRC elisp

(map! :map global-map
      :n "C-h" #'evil-prev-buffer
      :n "C-l" #'evil-next-buffer
      :n "C-j" #'evil-jump-forward
      :n "C-k" #'evil-jump-backward
      )

#+END_SRC

*** Comments

#+begin_src elisp

(map! :map global-map
      :nv ";" #'evilnc-comment-or-uncomment-lines)

#+end_src

** Windows

I find it rather handy to be asked which buffer I want to see after
splitting the window.

#+begin_src elisp

(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))

#+end_src

* Git (MaGit)

#+begin_src elisp

(print "[config.org] Git")

(map! :localleader
      :mode git-commit-mode
      :n :desc "Commit" "," #'with-editor-finish
      :n :desc "Quit commit" "q" #'with-editor-cancel)

(map! :leader
      (:prefix-map ("g" . "git")
       :desc "Magit status" "s" #'magit-status
       :desc "Magit status here" "S"   #'magit-status-here
       :desc "Git stage hunk" "g"   #'git-gutter:stage-hunk
       :desc "Git stage file" "G"   #'magit-stage-file
      ))
#+end_src

* Org

#+BEGIN_SRC elisp

(setq org-directory "~/org/")

(map! :after org
      :map org-mode-map
      :localleader
      "e" nil
      (:prefix-map ("e" . "edit / eval / export")
       "e" #'eval-last-sexp
       "E" #'org-export-dispatch
       "s" #'org-edit-special
       ))

;; none of these works :-(
(map! :after org
      :map org-src-mode-map
      :localleader
      "," #'org-edit-src-exit)
(define-key org-src-mode-map (kbd ", ,") #'org-edit-src-exit)

#+END_SRC

* LSP

- https://emacs-lsp.github.io/lsp-mode/tutorials/how-to-turn-off/

#+BEGIN_SRC elisp

(print "[config.org] LSP")

(map! :localleader
      :mode lsp-mode
      :n "=" #'lsp-format-buffer)

(with-eval-after-load 'lsp-mode
  (print "[config.org] with-eval-after-load lsp-mode")
  (setq lsp-ui-imenu-enable t
        lsp-ui-doc-enable t
        lsp-ui-sideline-show-code-actions nil
        lsp-lens-enable t
        lsp-enable-symbol-highlighting nil
        ;; lsp-enable-file-watchers nil
        ;; +lsp-prompt-to-install-server 'quiet
        )
  (add-to-list 'lsp-file-watch-ignored-directories "[/\\\\]\\firebase\\'")
  )

#+END_SRC

* LISP

#+BEGIN_SRC elisp

(after! lisp-mode
  (modify-syntax-entry ?- "w" lisp-mode-syntax-table))


;; (map! :localleader
      ;; ",a"  #'evil-cp-insert-at-end-of-form
      ;; ",i" 'evil-cp-insert-at-beginning-of-form
      ;; "(" #'sp-wrap-round)

#+END_SRC

* Clojure

#+BEGIN_SRC elisp

(print "[config.org] Clojure")

(after! clojure-mode
  (modify-syntax-entry ?- "w" clojure-mode-syntax-table))

(after! clojurescript-mode
  (modify-syntax-entry ?- "w" clojure-mode-syntax-table))

(after! clojurec-mode
  (modify-syntax-entry ?- "w" clojure-mode-syntax-table))

(map! :localleader
      :mode clojure-mode
      ;; "==" 'lsp-format-buffer
      ;; "(" 'sp-wrap-round
      ;; "#" 'cider-toggle-ignore-next-form
      "ev" #'cider-eval-sexp-at-point)

(map! :localleader
      :mode clojurescript-mode
      ",a" 'evil-cp-insert-at-end-of-form
      ",i" 'evil-cp-insert-at-beginning-of-form
      "ev" #'cider-eval-sexp-at-point)

(map! :localleader
      :mode clojurec-mode
      "ev" #'cider-eval-sexp-at-point)

#+END_SRC


doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;;           DO NOT MODIFY!
;;       tangeled from config.org

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
(setq user-full-name "Witoslaw Koczewski"
      user-mail-address "wi@koczewski.de")

(setq doom-localleader-key ",")

;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))


;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(print "[config.org] Auto-customizations")

(setq-default custom-file (expand-file-name "custom.el" doom-private-dir))
(when (file-exists-p custom-file)
  (load custom-file))

(setq-default
 ;; ad-redefinition-action 'accept         ; Silence warnings for redefinition
 auto-save-list-file-prefix nil         ; Prevent tracking for auto-saves
 cursor-in-non-selected-windows nil     ; Hide the cursor in inactive windows
 cursor-type '(hbar . 2)                ; Underline-shaped cursor
 custom-unlispify-menu-entries nil      ; Prefer kebab-case for titles
 custom-unlispify-tag-names nil         ; Prefer kebab-case for symbols
 delete-by-moving-to-trash t            ; Delete files to trash
 fill-column 80                         ; Set width for automatic line breaks
 gc-cons-threshold (* 8 1024 1024)      ; We're not using Game Boys anymore
 help-window-select t                   ; Focus new help windows when opened
 indent-tabs-mode nil                   ; Stop using tabs to indent
 inhibit-startup-screen t               ; Disable start-up screen
 initial-scratch-message ""             ; Empty the initial *scratch* buffer
 mouse-yank-at-point t                  ; Yank at point rather than pointer
 read-process-output-max (* 1024 1024)  ; Increase read size per process
 recenter-positions '(5 top bottom)     ; Set re-centering positions
 scroll-conservatively 101              ; Avoid recentering when scrolling far
 scroll-margin 12                       ; Add a margin when scrolling vertically
 select-enable-clipboard t              ; Merge system's and Emacs' clipboard
 scroll-preserve-screen-position 't
 sentence-end-double-space nil          ; Use a single space after dots
 ;; show-help-function nil                 ; Disable help text everywhere
 tab-always-indent 'complete            ; Tab indents first then tries completions
 tab-width 2                            ; Smaller width for tab characters
 uniquify-buffer-name-style 'forward    ; Uniquify buffer names
 ;; warning-minimum-level :error           ; Skip warning buffers
 window-combination-resize t            ; Resize windows proportionally
 x-stretch-cursor t)                    ; Stretch cursor to the glyph width

;; (blink-cursor-mode 0)                   ; Prefer a still cursor
(delete-selection-mode 1)               ; Replace region when inserting text
(fset 'yes-or-no-p 'y-or-n-p)           ; Replace yes/no prompts with y/n
(global-subword-mode 1)                 ; Iterate through CamelCase words
(mouse-avoidance-mode 'exile)           ; Avoid collision of mouse with point
(put 'downcase-region 'disabled nil)    ; Enable downcase-region
(put 'upcase-region 'disabled nil)      ; Enable upcase-region
(set-default-coding-systems 'utf-8)     ; Default to utf-8 encoding

;; Change a few indenting behaviors.
(put 'add-function 'lisp-indent-function 2)
(put 'advice-add 'lisp-indent-function 2)
(put 'plist-put 'lisp-indent-function 2)

(setq display-line-numbers-type nil)

(setq doom-theme 'doom-one)
;; (setq doom-theme 'doom-one-light)
;; (setq doom-theme 'doom-vibrant)
;; (setq doom-theme 'doom-gruvbox)
;; (setq doom-theme 'doom-dracula)

;; test ->

(setq doom-font (font-spec :family "Fira Code" :size 14)
      doom-variable-pitch-font (font-spec :family "Ubuntu" :size 15)
      ivy-posframe-font (font-spec :family "Fira Code" :size 17))

(setq confirm-kill-emacs nil)

(add-hook 'smartparens-enabled-hook #'smartparens-strict-mode)
(add-hook 'smartparens-enabled-hook #'evil-smartparens-mode)
(add-hook 'smartparens-enabled-hook #'evil-cleverparens-mode)

(map! :map global-map
      :mode evil-cleverparens-mode
      :n "M-l" #'evil-cp->)
(map! :map global-map
      :mode evil-cleverparens-mode
      :n "M-h" #'evil-cp-<)

(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

(map! :leader
      :desc "List commands"
      "SPC"  #'execute-extended-command)

(map! :map global-map
      :n "C-h" #'evil-prev-buffer
      :n "C-l" #'evil-next-buffer
      :n "C-j" #'evil-jump-forward
      :n "C-k" #'evil-jump-backward
      )

(map! :map global-map
      :nv ";" #'evilnc-comment-or-uncomment-lines)

(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))

(print "[config.org] Git")

(map! :localleader
      :mode git-commit-mode
      :n :desc "Commit" "," #'with-editor-finish
      :n :desc "Quit commit" "q" #'with-editor-cancel)

(map! :leader
      (:prefix-map ("g" . "git")
       :desc "Magit status" "s" #'magit-status
       :desc "Magit status here" "S"   #'magit-status-here
       :desc "Git stage hunk" "g"   #'git-gutter:stage-hunk
       :desc "Git stage file" "G"   #'magit-stage-file
      ))

(setq org-directory "~/org/")

(map! :after org
      :map org-mode-map
      :localleader
      "e" nil
      (:prefix-map ("e" . "edit / eval / export")
       "e" #'eval-last-sexp
       "E" #'org-export-dispatch
       "s" #'org-edit-special
       ))

;; none of these works :-(
(map! :after org
      :map org-src-mode-map
      :localleader
      "," #'org-edit-src-exit)
(define-key org-src-mode-map (kbd ", ,") #'org-edit-src-exit)

(print "[config.org] LSP")

(map! :localleader
      :mode lsp-mode
      :n "=" #'lsp-format-buffer)

(with-eval-after-load 'lsp-mode
  (print "[config.org] with-eval-after-load lsp-mode")
  (setq lsp-ui-imenu-enable t
        lsp-ui-doc-enable t
        lsp-ui-sideline-show-code-actions nil
        lsp-lens-enable t
        lsp-enable-symbol-highlighting nil
        ;; lsp-enable-file-watchers nil
        ;; +lsp-prompt-to-install-server 'quiet
        )
  (add-to-list 'lsp-file-watch-ignored-directories "[/\\\\]\\firebase\\'")
  )

(after! lisp-mode
  (modify-syntax-entry ?- "w" lisp-mode-syntax-table))


;; (map! :localleader
      ;; ",a"  #'evil-cp-insert-at-end-of-form
      ;; ",i" 'evil-cp-insert-at-beginning-of-form
      ;; "(" #'sp-wrap-round)

(print "[config.org] Clojure")

(after! clojure-mode
  (modify-syntax-entry ?- "w" clojure-mode-syntax-table))

(after! clojurescript-mode
  (modify-syntax-entry ?- "w" clojure-mode-syntax-table))

(after! clojurec-mode
  (modify-syntax-entry ?- "w" clojure-mode-syntax-table))

(map! :localleader
      :mode clojure-mode
      ;; "==" 'lsp-format-buffer
      ;; "(" 'sp-wrap-round
      ;; "#" 'cider-toggle-ignore-next-form
      "ev" #'cider-eval-sexp-at-point)

(map! :localleader
      :mode clojurescript-mode
      ",a" 'evil-cp-insert-at-end-of-form
      ",i" 'evil-cp-insert-at-beginning-of-form
      "ev" #'cider-eval-sexp-at-point)

(map! :localleader
      :mode clojurec-mode
      "ev" #'cider-eval-sexp-at-point)
